product_category_code_unique/models/product_category.py

```diff
@@ -1,16 +1,67 @@
 # Copyright 2021 ACSONE SA/NV
+# Copyright 2024 Manuel Regidor <manuel.regidor@sygel.es>
 # License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
 
-from odoo import api, models
+from odoo import _, api, models
+from odoo.exceptions import ValidationError
 
 
 class ProductCategory(models.Model):
 
     _inherit = "product.category"
 
-    _sql_constraints = [
-        ("uniq_code", "unique(code)", "The category code must be unique!"),
-    ]
+    def _get_parents(self):
+        self.ensure_one()
+        parent = self.parent_id
+        parents = self.env["product.category"]
+        while parent:
+            parents += parent
+            parent = parent.parent_id
+        return parents
+
+    def _get_children(self):
+        self.ensure_one()
+        return self.search([("parent_id", "child_of", self.id)])
+
+    def _get_hierarchy_cats(self):
+        self.ensure_one()
+        return self + self._get_parents() + self._get_children()
+
+    def _code_restriction(self, restriction=False):
+        restriction = restriction or self.env["ir.config_parameter"].get_param(
+            "product_code_unique.product_code_unique_restriction"
+        )
+        if restriction:
+            for cat in self:
+                domain = []
+                within = ""
+                if restriction == "system":
+                    domain = [("code", "=", cat.code)]
+                    within = _("the system")
+                elif restriction == "direct":
+                    to_check = cat.parent_id + cat.child_id
+                    domain = [
+                        "|",
+                        ("id", "=", cat.id),
+                        "&",
+                        ("code", "=", cat.code),
+                        ("id", "in", to_check.ids),
+                    ]
+                    within = _("parent and children")
+                elif restriction == "hierarchy":
+                    to_check = cat._get_hierarchy_cats()
+                    domain = [("code", "=", cat.code), ("id", "in", to_check.ids)]
+                    within = _("category hierarchy")
+                if self.search_count(domain) > 1:
+                    raise ValidationError(
+                        _("The category code must be unique within {within}!").format(
+                            within=within
+                        )
+                    )
+
+    @api.constrains("code")
```

Comment:
What if we have 2 categories with the same code but different parent, and then we edit the parent of one of the categories to be the same?

I think parent_id and child_ids fields should be added to constrains

Reply:
You're right. Changed.